### DISCUSSION TITLE
Turborepo integration with rushjs monorepo

### DISCUSSION THREAD

**Discussion by thanasis00** [vercel/turborepo#301] [Help] [answered]
### Describe the feature you'd like to request

Hello, in my team we maintain a [RushJS](https://rushjs.io/) monorepo, consisting of over 40 packages, and it really helped us move further from Lerna/Yarn monorepo setups. I found out about turborepo yesterday, and I wanted to try out some features (mostly the local and remote cache, and also the graph), but I find a hard time integrating it with our setup. RushJS uses pnpm under the hood, but it seems like turborepo uses specific hardlinks with specific files in order to integrate with a monorepo.

### Describe the solution you'd like

Turborepo could integrate with RushJS by configuring the location of the pnpm requirements (pnpm-workspace.yaml) and it could also either use a package.json only for its configuration, or use an external configuration file, like `turborepo.{json,yaml}` in order to provide the dependencies.

### Describe alternatives you've considered

RushJS is very fluid in terms of integrations, I could explain thoroughly a basic integration guide, if needed, until to the point of running turbo commands, which is uncharted territory from that point forward.

**Comment by jaredpalmer** [comment, marked as answer]
Turborepo works with pnpm workspaces. Rush and Turborepo are somewhat, if not directly, competitive tools. Although, Rush does more than just task running at the moment. 

The issue you'll face with moving from Rush are its rig files which avoid duplication of package.json content which turbo needs. 

To adopt turbo incrementally from rush, I believe you would need to re-establish all package.json's in their entirety. Then everything should work fine and turbo will be able to run tasks, even if rush is doing installation for the time being. The goal would be  to remove Rush entirely and simply rely on PNPM for installation instead. 

Hope that helps.

**Reply by thanasis00** [reply]
Thank you for the information. I understand that rush and turborepo seem competitive tools, but to be able to leverage as a user every tool's aspect which will aid you to maintain your project, could be useful. So I guess the best way would be to migrate to pnpm first, and then use turborepo for all tasks. Thank you for your time! I will keep an eye on this, as I find some turborepo's features really useful! Cheers!

**Reply by dlannoye** [reply]
There are several ways of using rush and not all of them involve using the rig files which is part of the larger [rush stack](https://rushstack.io/) project instead of a core feature of [rush](https://rushjs.io/). 

I currently have a rush monorepo that doesn't use the rig files, but instead looks very close to a standard monorepo using pnpm workspaces via rush. Beyond managing pnpm installs, rush is mainly used for orchestrating calls to our custom npm scripts. turborepo looks like it could be a good fit for us as we look at ways to keep scaling the repository. I would love to have a way to do incremental migration to turborep, but the main issue I see is getting turborepo to understand my packages while still having rush managing pnpm. 

Looking at the turbo repo code I see a couple ways something like this could be accommodated:
1. Add support for pnpm-workspace.yaml being in a user defined location instead of at the root. Using pnpm workspaces with rush, it keeps this file in ./common/temp/pnpm-workspaces.yaml and if I could specify that location I should be able to use both turborepo and rush side-by-side as I migrate.
2. Add a new [packagemanager](https://github.com/vercel/turborepo/tree/main/cli/internal/packagemanager) for rush. This would read the projects from [rush.json](https://rushjs.io/pages/configs/rush_json/). Without knowing too much about the internals of turborepo, I suspect this approach has more risk because rush can be used with a variety of package mangers so depending on how much you need to be able to assume about how node_modules are structured this might only work with a very specific configurations of rush and cause confusion for others trying to use turbo repo with a rush repo using features like rigs. 

Of these, it seems like 1. is really the best option. Is extra configuration like this something you ever see turborepo supporting?
